File: 5.Longest Palindromic Substring.cpp
/*
 * @lc app=leetcode id=5 lang=cpp
 * @lcpr version=20004
 *
 * [5] Longest Palindromic Substring
 *
 * https://leetcode.com/problems/longest-palindromic-substring/description/
 *
 * algorithms
 * Medium (35.01%)
 * Likes:    30153
 * Dislikes: 1852
 * Total Accepted:    3.5M
 * Total Submissions: 10.1M
 * Testcase Example:  '"babad"'
 *
 * Given a string s, return the longest palindromic substring in s.
 * 
 * 
 * Example 1:
 * 
 * Input: s = "babad"
 * Output: "bab"
 * Explanation: "aba" is also a valid answer.
 * 
 * 
 * Example 2:
 * 
 * Input: s = "cbbd"
 * Output: "bb"
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * 1 <= s.length <= 1000
 * s consist of only digits and English letters.
 * 
 * 
 */


// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start
class Solution {
public:
    string longestPalindrome(string s) {
        string res = "";
        for(int i = 0 ; i < s.size(); i++){
            string temp = check(s,i,i);
            string temp2 = check(s,i,i+1);
            cout<<temp<<" "<<temp2<<endl;
            if(temp.size()>res.size()) res = temp;
            if(temp2.size()>res.size()) res = temp2;
        }
        return res;
    }
    string check(string s, int l, int r){
        while(l >= 0 && r < s.size() && s[l] == s[r]){
            l--;
            r++; 
        }
        return s.substr(l+1,r-l-1);
    }
};
// @lc code=end



/*
// @lcpr case=start
// "babad"\n
// @lcpr case=end

// @lcpr case=start
// "cbbd"\n
// @lcpr case=end

 */

File: 5.Longest Palindromic Substring.py
#
# @lc app=leetcode id=5 lang=python3
# @lcpr version=20004
#
# [5] Longest Palindromic Substring
#
# https://leetcode.com/problems/longest-palindromic-substring/description/
#
# algorithms
# Medium (35.11%)
# Likes:    30245
# Dislikes: 1862
# Total Accepted:    3.6M
# Total Submissions: 10.2M
# Testcase Example:  '"babad"'
#
# Given a string s, return the longest palindromic substring in s.
# 
# 
# Example 1:
# 
# Input: s = "babad"
# Output: "bab"
# Explanation: "aba" is also a valid answer.
# 
# 
# Example 2:
# 
# Input: s = "cbbd"
# Output: "bb"
# 
# 
# 
# Constraints:
# 
# 
# 1 <= s.length <= 1000
# s consist of only digits and English letters.
# 
# 
#


# @lcpr-template-start

# @lcpr-template-end
# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        
        def isPan(s:str,i:int,j:int)->str:
            count = 0
            while i >= 0 and j < len(s) and s[i] == s[j]:
                i-=1
                j+=1
                
            return s[i+1:j]
        res = ""

        for i in range(len(s)):
            a = isPan(s,i,i) 
            b = isPan(s,i,i+1)
    
            if len(a) > len(res):
                res = a
            if len(b) > len(res):
                res = b
        return res
            
# @lc code=end



#
# @lcpr case=start
# "babad"\n
# @lcpr case=end

# @lcpr case=start
# "cbbd"\n
# @lcpr case=end

#

